### PR TITLE
Added all HTTP Status Codes

### PR DESCRIPTION
**Pull request by johannesschobel**
This PR adds **all** currently available http status codes (list taken from https://developer.mozilla.org/en-US/docs/Web/HTTP/Status ) to this project.

Please note, that the status code names have been changed accordingly - so this **MAY** be a breaking change for existing projects..

For example, the `Unavailable` has been renamed to `ServiceUnavailable`!

All the best

### CODE REVIEW

**Comment by jeffijoe** [comment]
I agree with all of them except 401, see this: https://stackoverflow.com/a/6937030/561545

Would you agree that keeping 401 as NotAuthenticated is a better idea?

**Comment by johannesschobel** [comment]
Why not adding BOTH status codes? and have the one as some kind of "alias" for the other one? I did the same with  `500` which is an `InternalServerError` **and** your predefined `GeneralError`.. Your opinion?

**Comment by jeffijoe** [comment]
So the reason I named it `GeneralError` and not `InternalServerError` was so that using them in non-HTTP contexts they wouldn't imply that a server is involved.

I suppose adding the aliases is fine, I believe you can do `export class InternalServerError extends GeneralError {}` and override the constructor in the alias to specify a different default message. You'd need to add tests for the aliases though. This would make `new InternalServerError() instanceof GeneralError` work, too.

Does that sound reasonable?

**Comment by johannesschobel** [comment]
Yeah, that would certainly be an option, but i guess it is easier to **not** extend them and just duplicate it..

Based on your answer, i am thinking about refactoring the predefined errors into something like
```
http\
   client.error.ts
   server.error.ts
general\
   general.error.ts
```

this way, we could have a better / easier to maintain structure.. 

Your thoughts?

**Comment by jeffijoe** [comment]
I think keeping them in the same file is fine. 

It's definitely easier to duplicate but you won't get the `instanceof` checks anymore. I'm honestly having second thoughts on the aliases because it's another thing to bikeshed and fragment over. 🤔

**Comment by johannesschobel** [comment]
hm, and what would you suggest?

**Comment by jeffijoe** [comment]
I'm not entirely sure; I definitely want all the HTTP codes to be supported but I'd like the existing ones to keep their names. The thing with multiple ways of doing the same thing in the same library, it causes people to argue over which one to use. If I in one place use `GeneralError` and you in another use `InternalServerError`, there's cognitive overhead in making out the fact that they actually mean the same thing.

Quoting the introductory comment in `http.ts`:

```
/**
 * Exposes common HTTP errors with a preconfigured `statusCode` for convenience.
 * This does not mean you have to use HTTP to use them, it's just that
 * HTTP has already defined a nice set of common error types, and tacking
 * on the HTTP status code makes it easier to consume in HTTP apps.
 */
```

The key bit being `This does not mean you have to use HTTP to use them`.

I suppose if we renamed `http.ts` to `common.ts` (it gets exported by the `index.ts` so it's not a breaking change), there'd be no expectation of the error names matching 1 to 1 with HTTP descriptions.

Thoughts?

**Comment by johannesschobel** [comment]
hmmmmmm.. not sure if i fully agree with this. Mostly, because: you say, that it can be used in a non-http context as well (comment), but you design the errors to be used in the http context (e.g., the status code, ...) and create them from the http response code blueprints.. 

If i would need to decide on one option we have discussed so far, i would go for the `InternalServerError extends GeneralError` approach.. I'm still thinking about splitting the errors into different files, so we could add different "packages" as well.. for example, one could add some `database` errors as well (e.g., a `NotFoundException` if the database cannot find requested items) or a `PreconditionFailed` exception, if no `id` is passed, or whatever.. 

Basically, the `database` package would contain a "subset" from the `http` package, with a different meaning.. however, they may be mapped to different `statusCode` values..

**Comment by jeffijoe** [comment]
The HTTP ones were added as convenience and because HTTP maps _pretty well_ (not perfect, as evidenced by `401 Unauthorized`) to common use cases.

For example, I often use `NotFound` in my data stores. The HTTP status code is present but I don't really use it until my HTTP error handler (if I am hosting it in a HTTP context) catches it.

> but you design the errors to be used in the http context and create them from the http response code blueprints.

Yes, for the codes and statuses that make sense in general. I used them as inspiration (and tacked on the status code for convenience) but I never made any guarantees about following a spec. 😄

**Comment by johannesschobel** [comment]
ok, and what would you suggest? i can rework my PR to better fit the use-cases of this package..

**Comment by jeffijoe** [comment]
I think to avoid breaking changes, let's keep the original `NotAuthenticated` and `GeneralError` and just add ones that are missing. You can leave the changes for `PaymentRequired` and `ServiceUnavailable` as I don't believe anybody is using them, the first 2 I mentioned are the most common ones.

Sounds reasonable?

**Comment by johannesschobel** [comment]
ok, i can do this.. don't know if i have time tomorrow..
so i will remove the `InternalServerError` as well?

All the best

**Comment by jeffijoe** [comment]
Yeah. 😀

No rush, take your time. 😄

**Comment by johannesschobel** [comment]
done.. just give me some feedback if you have time..
All the best!

**Comment by jeffijoe** [comment]
This looks good! To merge this, I need you to **rebase** to the tip of master, and **squash** your 2 commits into a single one with a message like `feat: add remaining HTTP status codes`. 😄

If you are on a Mac, I can recommend [Fork](https://git-fork.com), a really nice Git UI.

**Comment by johannesschobel** [comment]
holy shit.. i think i did it.. git console is weird.. ^^ can you please verify if this is ok?
All the best

**Comment by jeffijoe** [comment]
Excellent, thanks for the PR!

**Comment by johannesschobel** [comment]
thank you for the support and help.. glad i could contribute.. ^^
